### PR TITLE
Tuning guide for low latency and localization

### PR DESCRIPTION
**Pull request by gaschler**
Related to #397

### CODE REVIEW

File: docs/source/tuning.rst [hunk 1]
```diff
@@ -125,6 +125,60 @@ Experimenting with this value yields a better result at ``2e2``.
 Here, the scan matcher used rotation to still slightly mess up the result though.
 Setting the ``rotation_weight`` to ``4e2`` leaves us with a reasonable result.
 
+Special Cases
+-------------
+
+The default configuration and the above tuning steps are focused on quality.
+Only after we have achieved good quality, we can further consider low latency tasks and other special cases.
```
Comment (SirVer) [inline docs/source/tuning.rst]: Define 'low latency'

Comment (SirVer) [reply docs/source/tuning.rst]: Put this information first and mention you're talking about low latency later?

File: docs/source/tuning.rst [hunk 2]
```diff
@@ -125,6 +125,60 @@ Experimenting with this value yields a better result at ``2e2``.
...
+
+Low Latency
+^^^^^^^^^^^
+
+Low latency is required for online algorithms, such as robot localization.
+Local SLAM, which operates in the foreground, directly affects latency.
+Global SLAM builds up a queue of background tasks. When global SLAM cannot keep up the the queue,
```
Comment (SirVer) [inline docs/source/tuning.rst]: one sentence per line

Comment (SirVer) [inline docs/source/tuning.rst]: the the

File: docs/source/tuning.rst [hunk 3]
```diff
@@ -125,6 +125,60 @@ Experimenting with this value yields a better result at ``2e2``.
...
+drift can accumulate indefinitely, so global SLAM should be tuned to work in real time.
+
+To tune the different components for speed, there are many options, and we list them ordered from
```
Comment (SirVer) [inline docs/source/tuning.rst]: switch sentence around: There are many options to..., and we

File: docs/source/tuning.rst [hunk 4]
```diff
@@ -125,6 +125,60 @@ Experimenting with this value yields a better result at ``2e2``.
...
+Low Latency
+^^^^^^^^^^^
+
+Low latency is required for online algorithms, such as robot localization.
+Local SLAM, which operates in the foreground, directly affects latency.
+Global SLAM builds up a queue of background tasks. When global SLAM cannot keep up the the queue,
+drift can accumulate indefinitely, so global SLAM should be tuned to work in real time.
+
+To tune the different components for speed, there are many options, and we list them ordered from 
+the recommended, effective ones to the those that are more intrusive.
```
Comment (SirVer) [inline docs/source/tuning.rst]: what you mean with effective?

File: docs/source/tuning.rst [hunk 5]
```diff
@@ -125,6 +125,60 @@ Experimenting with this value yields a better result at ``2e2``.
...
+It is recommended to only explore one option at a time, starting with the first.
+
+To tune global SLAM for lower latency, we can
+
+- decrease ``optimize_every_n_nodes``
```
Comment (SirVer) [inline docs/source/tuning.rst]: Link to their documentations?

File: docs/source/tuning.rst [hunk 6]
```diff
@@ -125,6 +125,60 @@ Experimenting with this value yields a better result at ``2e2``.
...
+Low latency is required for online algorithms, such as robot localization.
+Local SLAM, which operates in the foreground, directly affects latency.
+Global SLAM builds up a queue of background tasks. When global SLAM cannot keep up the the queue,
+drift can accumulate indefinitely, so global SLAM should be tuned to work in real time.
+
+To tune the different components for speed, there are many options, and we list them ordered from 
+the recommended, effective ones to the those that are more intrusive.
+It is recommended to only explore one option at a time, starting with the first.
+
+To tune global SLAM for lower latency, we can
```
Comment (SirVer) [inline docs/source/tuning.rst]: Add a sentence here that explains that there is a certain threshold, that is not easy to determine: you can crank everything up as high as possible, which means you are slow but you have the best quality. The goal now is to go as low as possible without loosing quality, but gaining performance. You never want to do more work than this.

And if needed you can cross over and trade quality for performance.

File: docs/source/tuning.rst [hunk 7]
```diff
@@ -125,6 +125,60 @@ Experimenting with this value yields a better result at ``2e2``.
...
+- decrease ``max_range`` (especially if data is noisy)
+- decrease ``submaps.num_range_data``
+
+Note that larger voxels will slightly increase scan matching scores as a side effect,
+so score thresholds should be increased accordingly.
+
+Pure Localization in a Given Map
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Pure localization is different from mapping in that we expect a lower latency of both local and global
```
Comment (SirVer) [inline docs/source/tuning.rst]: Split runaway sentence into at least two.

File: docs/source/tuning.rst [hunk 8]
```diff
@@ -125,6 +125,60 @@ Experimenting with this value yields a better result at ``2e2``.
...
+It is recommended to only explore one option at a time, starting with the first.
+
+To tune global SLAM for lower latency, we can
+
+- decrease ``optimize_every_n_nodes``
+- increase ``MAP_BUILDER.num_background_threads`` up to the number of cores
+- decrease ``global_sampling_ratio``
+- decrease ``constraint_builder.sampling_ratio``
+- increase ``constraint_builder.min_score``
+- for the adaptive voxel filter(s), decrease ``.min_num_points``, ``.max_range``, increase ``.max_length``
```
Comment (SirVer) [inline docs/source/tuning.rst]: what about search radius?

Comment (gaschler) [reply docs/source/tuning.rst]: Done

File: docs/source/tuning.rst (hunk 1 above)
Comment (gaschler) [reply docs/source/tuning.rst]: Done

File: docs/source/tuning.rst (hunk 2 above)
Comment (gaschler) [reply docs/source/tuning.rst]: Done

Comment (gaschler) [reply docs/source/tuning.rst]: Done

File: docs/source/tuning.rst (hunk 3 above)
Comment (gaschler) [reply docs/source/tuning.rst]: Done

File: docs/source/tuning.rst (hunk 4 above)
Comment (gaschler) [reply docs/source/tuning.rst]: Corrected to "straightforward"

File: docs/source/tuning.rst (hunk 6 above)
Comment (gaschler) [reply docs/source/tuning.rst]: Done

File: docs/source/tuning.rst (hunk 5 above)
Comment (gaschler) [reply docs/source/tuning.rst]: Done

File: docs/source/tuning.rst (hunk 7 above)
Comment (gaschler) [reply docs/source/tuning.rst]: Done

**Comment by gaschler** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user gaschler. Merge queue now has a length of 1.

**Pull request merged**